### PR TITLE
[CI] Use LocalExecutor if kubeconfig is defined

### PR DESCRIPTION
**Pull request by nebril**
Executor interface is implemented by both LocalExecutor and
SSHMeta

This PR should not change the way tests are run on CI, but introduces code which will allow to switch to local kubectl execution as soon as we allow access to vagrant k8s cluster from host and are able to gather logs from all nodes via kubectl instead of ssh commands.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8683)

### CODE REVIEW

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Review by eloycoto — COMMENTED** [review]
This is super cool :-) Thanks!

**Comment by eloycoto** [comment]
Btw: Just a quick note, there are some places where helpers try to connect to all servers, I'm not sure if you try that!

https://github.com/cilium/cilium/blob/master/test/helpers/kubectl.go#L1914

**Comment by nebril** [comment]
@eloycoto yeah, I know that log gathering is still a problem, I have an idea on how to tackle this, that's why this PR is marked as WIP :). Thanks for the heads-up!

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
kubectl is currently provided in very hacky way, we need to wait for it to be available in CI nodes (pr pending)

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
k8s tests passed! runtime tests re-enabled

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

File: test/get-vagrant-kubeconfig.sh
Comment (ianvernon) [inline test/get-vagrant-kubeconfig.sh:1]: This have `-e` as a flag so that if it fails, provisioning fails.

File: ginkgo.Jenkinsfile
Comment (ianvernon) [inline ginkgo.Jenkinsfile]: Indentation is off.

File: ginkgo.Jenkinsfile
Comment (ianvernon) [inline ginkgo.Jenkinsfile]: Here as well.

File: ginkgo.Jenkinsfile
Comment (ianvernon) [inline ginkgo.Jenkinsfile]: Is it possible to put this all into a script? Would be a bit easier to read than having to do a bunch of `sh` commands in a row in the Jenkinsfile.

File: test/k8sT/manifests/log-gatherer.yaml
Comment (ianvernon) [inline test/k8sT/manifests/log-gatherer.yaml]: Create a TODO to have this be made to be part of the Cilium image repo.

File: test/provision/k8s_install.sh
Comment (ianvernon) [inline test/provision/k8s_install.sh]: Indentation looks off.

File: test/helpers/cons.go
Comment (ianvernon) [inline test/helpers/cons.go]: This comment is wrong - it is a relative path now.

File: test/k8sT/Policies.go
Comment (ianvernon) [inline test/k8sT/Policies.go]: Why not just have these all by `demoPath, l3Policy ... string`?

**Comment by nebril** [comment]
test-me-please

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: Why is this commented out?

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: Create a context with a reasonable timeout (e.g., 1 minute) so that the test will fail if the command gets stuck.

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: Why is this removed?

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: helpers.Kubesystem - perhaps hardcode the `k8s-app=cilium-test-logs` elsewhere as well - I want to avoid hardcoding of strings to make the code more maintainable.

Comment (ianvernon) [inline test/helpers/kubectl.go]: Use a context with a timeout here as well.

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: What is the rationale behind this loop? Is it because we only need one of the results at all?

File: test/helpers/kubectl.go
Comment (ianvernon) [inline test/helpers/kubectl.go]: Why is this commented out?

File: test/k8sT/Chaos.go
Comment (ianvernon) [inline test/k8sT/Chaos.go:235]: How does reducing the timeout relate to the killing of the process?

File: test/helpers/cons.go
Comment (ianvernon) [reply test/helpers/cons.go]: Fix this discrepancy.

File: test/helpers/cons.go
Comment (ianvernon) [reply test/helpers/cons.go]: Fix this discrepancy.

**Review by ianvernon — CHANGES_REQUESTED** [review]
Looks reasonable overall. I think this is a great first step. I have some misc. nits / feedback, but no big concerns.

File: test/helpers/kubectl.go
Comment (nebril) [reply test/helpers/kubectl.go]: Edited - I didn't know the purpose of this command, it's back now.

File: test/helpers/kubectl.go
Comment (nebril) [reply test/helpers/kubectl.go]: same as line 843

Comment (nebril) [reply test/helpers/kubectl.go]: Also, why is this necessary?

File: test/helpers/kubectl.go
Comment (nebril) [reply test/helpers/kubectl.go]: Changed it to different helper function which I created after this code.

File: test/helpers/kubectl.go
Comment (nebril) [reply test/helpers/kubectl.go]: It's back now.

File: test/k8sT/Chaos.go
Comment (nebril) [reply test/k8sT/Chaos.go:235]: Killing the process via `cancel` causes the netperf run via `os/exec` to return error, which causes the test to fail. I am not sure how to stop netperf non-destructively at this point, so I reduced the timeout and simply wait for it to finish. I agree that it's not ideal.

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by ianvernon** [comment]
Provisioning problems: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/14884/

**Comment by ianvernon** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
resolved conflict, rebased, previous run succeeded

**Comment by nebril** [comment]
test-me-please

File: test/helpers/kubectl.go
Comment (raybejjani) [inline test/helpers/kubectl.go:119]: Since `config.CiliumTestConfig.Kubeconfig` is the only global here, maybe just pass it in? When we come to change this (again) we'll have 1 less thing to think about.

File: test/k8sT/Chaos.go
Comment (raybejjani) [inline test/k8sT/Chaos.go:236]: Unless the double cancel was causing an issue, it would be better to leave it out for another fixups PR.

File: test/k8sT/KafkaPolicies.go
Comment (raybejjani) [inline test/k8sT/KafkaPolicies.go]: Maybe comment that these are set in BeforeAll below. It's more than a page away and may not be obvious.

File: test/k8sT/manifests/log-gatherer.yaml
Comment (raybejjani) [inline test/k8sT/manifests/log-gatherer.yaml:26]: I imagine the image is pretty stable now, so we could move it before we merge if you want.

File: test/test_suite_test.go
Comment (raybejjani) [inline test/test_suite_test.go]: I'm pretty sure `kubectl :=` is the same as `var kubectl *helpers.Kubectl`. Was the change just to make something more obvious?

File: test/vagrant-ci-start.sh
Comment (raybejjani) [inline test/vagrant-ci-start.sh:15]: Perhaps having an echo explaining what this line is for will help when it breaks. The rest of the steps may also be improved with some logging.

File: test/vagrant-ci-start.sh
Comment (raybejjani) [inline test/vagrant-ci-start.sh]: Is this just for debugging so it's in the logs somewhere? If so, wouldn't `cat vagrant-kubeconfig` be more correct?

File: test/helpers/cilium.go
Comment (raybejjani) [inline test/helpers/cilium.go:339]: The commit message for `[CI] Move BaseString to SSHMeta and Kubectl structs` is wrong, since it's BasePath. Also, since it's a more general change, could you add some explanation in the commit ?

File: test/helpers/kubectl.go
Comment (raybejjani) [inline test/helpers/kubectl.go]: Since we have `.BasePath()` on the SSHMeta type, maybe it would be easier to add that to Executor and use that where we use this field?

**Review by raybejjani — CHANGES_REQUESTED** [review]

File: test/helpers/cilium.go
Comment (nebril) [reply test/helpers/cilium.go:339]: done

File: test/helpers/kubectl.go
Comment (nebril) [reply test/helpers/kubectl.go]: done

File: test/helpers/kubectl.go
Comment (nebril) [reply test/helpers/kubectl.go:119]: Not sure about that, we don't leak config to tests themselves right now, and this change would need for tests to be aware of user config. If you insist I can make this change.

File: test/k8sT/Chaos.go
Comment (nebril) [reply test/k8sT/Chaos.go:236]: This cancel was killing the kubectl process as opposed to closing the ssh connection (previous version). 

As I wrote above: Killing the process via cancel causes the `kubectl exec netperf` run via os/exec to return error, which causes the test to fail. I am not sure how to stop netperf non-destructively at this point, so I reduced the timeout and simply wait for it to finish. I agree that it's not ideal.

File: test/vagrant-ci-start.sh
Comment (nebril) [reply test/vagrant-ci-start.sh]: `cat vagrant-kubeconfig` will give already processed kubeconfig, this was added to show what is actually in source file. I can delete this, as it was for debugging.

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
I forgot to "submit review" so Ian's comments were unsanswered. Sorry!

File: test/k8sT/manifests/log-gatherer.yaml
Comment (nebril) [reply test/k8sT/manifests/log-gatherer.yaml:26]: Don't have access rights to our dockerhub (security!). Will change this in followup PR

Comment (raybejjani) [reply test/k8sT/manifests/log-gatherer.yaml:26]: neither do I! We'll do it later.

File: test/helpers/kubectl.go
Comment (raybejjani) [reply test/helpers/kubectl.go:119]: But by doing this we are leaking, no? The import tree now includes the config package. It's not that important, though.

**Review by raybejjani — APPROVED** [review]

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-missed-k8s

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-missed-k8s

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-missed-k8s

**Review by ianvernon — APPROVED** [review]

File: test/helpers/kubectl.go
Comment (raybejjani) [inline test/helpers/kubectl.go]: @nebril I noticed that we were using .Join but we already have a `GetFilePath` that does it for us on the kubectl type.

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-missed-k8s

**Comment by nebril** [comment]
squashed commits

**Comment by mrostecki** [comment]
Backport is needed for 1.6 as a dependency for #9342

**Comment by ianvernon** [comment]
Removing backport tag per discussion on #9342

**Pull request merged**
